File: src/main/java/live/adamlearns/animalroyale/GamePlayer.java
package live.adamlearns.animalroyale;

import org.bukkit.ChatColor;
import org.bukkit.DyeColor;
import org.bukkit.entity.Sheep;

/**
 * This is NOT the same as a Minecraft Player (i.e. an org.bukkit.entity.Player). This represents someone from Twitch
 * chat who has typed commands to play the game.
 */
public class GamePlayer {
    /**
     * This is the person's Twitch name.
     */
    private final String name;
    private boolean hasAddedSheep = false;
    private Sheep sheep;

    /**
     * This is as close to your sheep's color as we can get. We approximate some of them like black so that it's still
     * readable.
     */
    private ChatColor nameColor;

    private int tntNextYaw = 0;
    private int tntNextPitch = 0;
    private int tntNextPower = 0;
    private double tntNextTtl = 0;
    private boolean hasSetTntParameters = false;

    // This is essentially measured in System.currentTimeMillis
    private long nextTimeAbleToUseSpecialAbility = 0;

    public static final int MIN_PITCH = -90;
    public static final int MAX_PITCH = 90;
    public static final int MIN_POWER = 0;
    public static final int MAX_POWER = 100;
    public static final double MIN_TTL = 0;
    public static final double MAX_TTL = 5.0;

    public GamePlayer(final String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setTntParameters(final int yaw, final int pitch, final int power, final double ttl) {
        setYaw(yaw);
        setPitch(pitch);
        setPower(power);
        setTtl(ttl);
        hasSetTntParameters = true;
    }

    public void setYaw(final int yaw) {
        tntNextYaw = yaw;
    }

    public void setPitch(final int pitch) {
        tntNextPitch = Util.clamp(pitch, MIN_PITCH, MAX_PITCH);
    }

    public void setPower(final int power) {
        tntNextPower = Util.clamp(power, MIN_POWER, MAX_POWER);
    }

    public void setTtl(final double ttl) {
        tntNextTtl = Util.clamp(ttl, MIN_TTL, MAX_TTL);
    }

    public void addYaw(final int yaw) {
        setYaw(tntNextYaw + yaw);
    }

    public void addPitch(final int pitch) {
        setPitch(tntNextPitch + pitch);
    }

    public void addPower(final int power) {
        setPower(tntNextPower + power);
    }

    public void addTtl(final double ttl) {
        setTtl(tntNextTtl + ttl);
    }

    public void setSheep(final Sheep sheep) {
        this.sheep = sheep;
        hasAddedSheep = true;

        final DyeColor sheepColor = sheep.getColor();

        assert sheepColor != null;
        nameColor = GamePlayer.getChatColorFromDyeColor(sheepColor);
    }

    public String getNameColoredForInGameChat() {
        return nameColor + name;
    }

    public String getNameForScoreboardWhenDead() {
        // The order for coloring a struck-out name is COLOR + STRIKETHROUGH, not the other way around.
        return nameColor.toString() + ChatColor.STRIKETHROUGH + name;
    }

    static ChatColor getChatColorFromDyeColor(final DyeColor dyeColor) {
        switch (dyeColor) {
            case BLACK:
            case LIGHT_GRAY:
                return ChatColor.GRAY;
            case BROWN:
            case ORANGE:
                return ChatColor.GOLD;
            case CYAN:
            case LIGHT_BLUE:
                return ChatColor.AQUA;
            case LIME:
                return ChatColor.GREEN;
            case GREEN:
                return ChatColor.DARK_GREEN;
            case MAGENTA:
            case PINK:
                return ChatColor.LIGHT_PURPLE;
            case PURPLE:
                return ChatColor.DARK_PURPLE;
            default:
                return ChatColor.valueOf(dyeColor.name());
        }
    }

    public boolean canPlaceSheep() {
        return !hasAddedSheep;
    }

    public boolean hasAddedSheep() {
        return hasAddedSheep;
    }

    public Sheep getSheep() {
        return sheep;
    }

    public boolean isSheepAlive() {
        return hasAddedSheep && sheep.isValid() && !sheep.isDead();
    }

    public int getTntNextYaw() {
        return tntNextYaw;
    }

    public int getTntNextPitch() {
        return tntNextPitch;
    }

    public int getTntNextPower() {
        return tntNextPower;
    }

    public double getTntNextTtl() {
        return tntNextTtl;
    }

    public boolean hasSetTntParameters() {
        return hasSetTntParameters;
    }

    public boolean canUseSpecialAbility() {
        return System.currentTimeMillis() >= nextTimeAbleToUseSpecialAbility;
    }

    public void setNextTimeAbleToUseSpecialAbility(final long nextTimeAbleToUseSpecialAbility) {
        this.nextTimeAbleToUseSpecialAbility = nextTimeAbleToUseSpecialAbility;
    }

    public String getNameForTwitch() {
        return "@" + name;
    }

    public String getCooldownMessage() {
        final float secRemaining = (nextTimeAbleToUseSpecialAbility - System.currentTimeMillis()) / 1000.0f;
        return String.format("Your ability is on cooldown for another %.2fs", secRemaining);
    }
}

File: src/main/kotlin/live/adamlearns/animalroyale/GamePlayer.kt
package live.adamlearns.animalroyale

import live.adamlearns.animalroyale.extensions.asTextColor
import live.adamlearns.animalroyale.extensions.clamp
import live.adamlearns.animalroyale.extensions.isAliveAndValid
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.text.format.TextDecoration
import org.bukkit.entity.Sheep

/**
 * This is NOT the same as a Minecraft Player (i.e. an org.bukkit.entity.Player). This represents someone from Twitch
 * chat who has typed commands to play the game.
 */
class GamePlayer(
    /**
     * This is the person's Twitch name.
     */
    val name: String,
    private val displayName: String?
) {
    /**
     * This is as close to your sheep's color as we can get. We approximate some of them like black so that it's still
     * readable.
     */
    private var nameColor: TextColor? = null
    // This is essentially measured in System.currentTimeMillis
    private var nextTimeAbleToUseSpecialAbility: Long = 0

    var sheep: Sheep? = null
        set(value) {
            field = value
            nameColor = sheep?.color?.asTextColor()
        }

    val canPlaceSheep: Boolean
        get() = sheep == null

    val hasAddedSheep: Boolean
        get() = sheep != null

    val isSheepAlive: Boolean
        get() = sheep?.isAliveAndValid ?: false

    var hasSetTntParameters = false
        private set
    var tntNextYaw: Int = 0
        private set
    var tntNextPitch: Int = 0
        private set
    var tntNextPower: Int = 0
        private set
    var tntNextTtl: Double = 0.0
        private set

    val colorfulName: TextComponent
        get() = Component.text(displayName ?: name, nameColor)

    val nameForScoreboardWhenDead: TextComponent
        get() = Component.text(displayName ?: name).color(nameColor).decorate(TextDecoration.STRIKETHROUGH)

    val nameForTwitch: String
        get() = "@${displayName ?: name}"

    val cooldownMessage: String
        get() {
            val secRemaining =
                (nextTimeAbleToUseSpecialAbility - System.currentTimeMillis()) / 1000.0f
            return String.format("Your ability is on cooldown for another %.2fs", secRemaining)
        }

    fun setTntParameters(yaw: Int, pitch: Int, power: Int, ttl: Double) {
        setYaw(yaw)
        setPitch(pitch)
        setPower(power)
        setTtl(ttl)
        hasSetTntParameters = true
    }

    fun addYaw(yaw: Int) {
        setYaw(tntNextYaw + yaw)
    }

    fun addPitch(pitch: Int) {
        setPitch(tntNextPitch + pitch)
    }

    fun addPower(power: Int) {
        setPower(tntNextPower + power)
    }

    fun addTtl(ttl: Double) {
        setTtl(tntNextTtl + ttl)
    }

    fun canUseSpecialAbility(): Boolean {
        return System.currentTimeMillis() >= nextTimeAbleToUseSpecialAbility
    }

    fun setNextTimeAbleToUseSpecialAbility(nextTimeAbleToUseSpecialAbility: Long) {
        this.nextTimeAbleToUseSpecialAbility = nextTimeAbleToUseSpecialAbility
    }

    fun temporarilyStopShootingTnt() {
        hasSetTntParameters = false
    }

    private fun setYaw(yaw: Int) {
        tntNextYaw = yaw
    }

    private fun setPitch(pitch: Int) {
        tntNextPitch = pitch.clamp(MIN_PITCH, MAX_PITCH)
    }

    private fun setPower(power: Int) {
        tntNextPower = power.clamp(MIN_POWER, MAX_POWER)
    }

    private fun setTtl(ttl: Double) {
        tntNextTtl = ttl.clamp(MIN_TTL, MAX_TTL)
    }

    companion object {
        const val MIN_PITCH: Int = -90
        const val MAX_PITCH: Int = 90
        const val MIN_POWER: Int = 0
        const val MAX_POWER: Int = 100
        const val MIN_TTL: Double = 0.0
        const val MAX_TTL: Double = 5.0
    }
}
